### DISCUSSION TITLE
Resource not accessible by personal access token for dispatch of workflow in project repository

### DISCUSSION THREAD

**Discussion by 3ng7n33r** [community/community#58868] [Apps, API and Webhooks] [answered]
### Select Topic Area

Question

### Body

I want to trigger a workflow from a repository dispatch within the repository of an organization.
I am sending a POST request to:
https://api.github.com/repos/_project_/_repo_/dispatches
with a personal access token in the authorization header like explained [here](https://docs.github.com/en/rest/repos/repos?apiVersion=2022-11-28#create-a-repository-dispatch-event)

It is a fine grained access token with the organization as owner and the following rights:

- Read access to code and metadata
- Read and Write access to actions 

When I send the request I get the following 403 response: 

{
    "message": "Resource not accessible by personal access token",
    "documentation_url": "https://docs.github.com/rest/reference/repos#create-a-repository-dispatch-event"
}

I am not sure what I am supposed to use. How else can I authenticate for repository dispatch events of an organization?
The "Allow access via fine-grained personal access tokens" setting of that organization is activated so that's not it.

**Comment by swati-tc** [comment]
Did you find any solution for this? I'm running into a similar issue.

**Reply by zoul** [reply]
We had a similar problem and what helped us was regenerating the token.

**Reply by kevholmes** [reply, marked as answer]
I tried re-generating with no luck. I was trying to make a PUT request to create a new environment (the fine-grained token had read+write access to the `Environments` scope which is described as `Manage repository environments`) and kept getting a 403. I eventually had to give the fine-grained token read+write to `Administration` which is described as `Repository creation, deletion, settings, teams, and collaborators.` and the request successfully goes through after that modification only 🤷🏽

**Reply by gberenice** [reply]
We have a similar issue with a fine-grained PAT created by a Member of Organization (requested and approved). 
It's a request [GET /repos/{owner}/{repo}/vulnerability-alerts](https://docs.github.com/en/rest/repos/repos#check-if-vulnerability-alerts-are-enabled-for-a-repository), which according to the docs require `read` for Repository administration, and these permissions are granted.
If a PAT is created by an Owner with the same list of permissions, all good.

**Reply by Qix-** [reply]
This is silly. This is giving a backdoor to any repository given a leak instead of scoping it directly to a specific action.

It's been 10 years, Github. Get your ACLs figured out already.

**Comment by rwojsznis** [comment]
If you are coming here from _the internet search_ - do yourself a favor and check _repository access_ as well when using scoped tokens - kinda already covered by @gberenice, but lemme elaborate:

I had a token with `read & write` access to the `actions` and `read` to `metadata`; triggering workflow should _just work_, right? Wrong. (Bot account) User had a `triage` access to the repository and it seems what's needed is a `write` **repo access** to trigger the workflow. 

The 403 I was getting was really confusing and I'm falling into same trap _every freaking time_ I had to create a new scoped token 😅

**Reply by tanguySnoeck** [reply]
Such a nasty issue, this did it for me. If the PAT permissions follow the documentation but you still hit a 403, the problem comes from the permission of the user in that repo. Would be nice to have this added to the documentation of the REST API.

**Comment by alejoar** [comment]
The fine-grained permission you need to trigger an action workflow is "Contents".

**Reply by apoorv-deriv** [reply]
Contents -> Read-Only with Actions -> Read and write worked for me.

**Reply by hoa-pham82** [reply]
Thank you so much, that saved my life!!

**Reply by PaiJi** [reply]
Thanks!

**Reply by david-alvez** [reply]
Thanks!

**Reply by kountaydwivedi** [reply]
> The fine-grained permission you need to trigger an action workflow is "Contents".

This helped me. thanks.

**Reply by Foticing** [reply]
Thank you for your help. Your assistance successfully resolved a problem that had been unsolved for four hours.

**Comment by juristr** [comment]
What I needed to set is `read/write` on the Contents permissions: https://docs.github.com/en/rest/authentication/permissions-required-for-fine-grained-personal-access-tokens?apiVersion=2022-11-28#repository-permissions-for-contents

**Reply by arashmad** [reply]
That's a cool documentation that shows right access for each action.
I tried to sync labels from a `.yml` configuration file and I got `403` using a custom action.
Then by checking this link you mentioned, I find only `read` and `write` in issues are enough.

**Comment by oerol** [comment]
I was able to dispatch a workflow by setting the `Actions` permission to `Read and write` in my fine-grained personal access token. Setting the `Contents` permission did not help me here.

**Comment by Tyr-Wang** [comment]
Plz watch out workflow type that you want to trigger: if you want to call `/repo/:owner/:repo/dispatch` which means `repo dispatch`, then u must set `Contents:write`. see manual [here](https://docs.github.com/en/rest/repos/repos?apiVersion=2022-11-28#create-a-repository-dispatch-event--fine-grained-access-tokens). It works without `actions` permission.

but if you want to trigger `/repos/{owner}/{repo}/actions/workflows/{workflow_id}/dispatches` which means `workflow dispatch` (equals manual trigger), then u need `actions:write`. see manual [here](https://docs.github.com/en/rest/actions/workflows?apiVersion=2022-11-28#create-a-workflow-dispatch-event--fine-grained-access-tokens)

**Reply by actforjason** [reply]
This solved my problem

**Comment by petabook** [comment]
As of April 2025, I had success with: 
Read access to metadata
Read and Write access to actions 
Read and Write access to workflows

**Comment by JohnHammerSky** [comment]
I was struggling with this issue because I didn't notice the drop-down "Access: Read-only" just in the right size of the permissions name and description.

The "Add permission" button opens another drop-down that hides (at first) this button, making it harder to spot.

I got it fixed changing the permission I chose to "Read and write" instead of "Read-only", which is the default.
